tmt/steps/prepare/__init__.py

```diff
@@ -86,58 +91,131 @@ def how_generic(self, how, what):
 
-            self.debug('Looking for prepare script', whatpath)
             if os.path.exists(whatpath) and os.path.isfile(whatpath):
                 what = whatpath
 
+            elif how == 'shell':
+                return self.command(what)
+
         try:
             self.plan.provision.prepare(how, what)
         except AttributeError as error:
             raise SpecificationError('NYI: cannot currently run this preparator.')
 
+    def how_install(self, how, what):
+        """ Install packages
+            handles various URIs or packages themselves
+        """
+        if not type(what) is list:
+            return self.install(what)
+
+        rest = []
+        for wha in what:
+            if self.is_uri(wha):
+                domain = self.get_uri(wha).netloc
+                if not re.search(r"^koji\.", domain) is None:
+                    self.how_koji('koji', how, what)
+                    continue
+                if not re.search(r"^brew\.", domain) is None:
+                    self.how_koji('brew', how, what)
+                    continue
+
+            rest += wha
+        self.install(rest)
+
+    def how_koji(self, com, how, what):
+        """ Download and install packages from koji URI """
+        what_uri = self.get_uri(what)
+
+        if what_uri:
+            query = self.get_query(what_uri)
+
+            if not 'buildID' in query:
+                raise SpecificationError(f"No buildID found in: {what}")
+
+            build = query['buildID']
+
+        else:
+            self.debug(f"Could not parse URI, assuming buildID was given.")
+            build = what
+
+        self.install(com)
+
+        install_dir = os.path.join(self.workdir, 'install')
+        self.command(f"{self.cmd_mkcd(install_dir)}; {com} download-build -a noarch -a x86_64 {build}; dnf install --skip-broken -y *.rpm; rm *.rpm")
+
 
     ## Additional API ##
+    def copr(self, copr):
+        """ Enable copr repository """
+        self.debug(f'Enabling copr repository', copr)
+        # Shouldn't be needed
+        # self.install('dnf-command(copr)')
+        return self.command(f"sudo dnf copr list --enabled | grep -qE '(^|\/){copr}$' || sudo dnf copr -y enable {copr}")
+
     def install(self, packages):
         """ Install specified package(s)
         """
         if type(packages) is list:
+            packages = [
+                self.quote(package) for package in packages
+            ]
             packages = ' '.join(packages)
+        else:
+            packages = self.quote(packages)
 
-        ## TODO: remove this after run(shell=True) is in provision.prepare()
-        try:
-            self.plan.provision.prepare('shell', f"rpm -V {packages}")
-        except GeneralError:
-            self.plan.provision.prepare('shell', f"dnf install -y {packages}")
-        return
-        ## <
+        return self.command(f"rpm -V {packages} || sudo dnf install -y {packages}")
 
+    def command(self, command, logfile=True):
```

Comment:
Is it really necessary to store the command in an extra script and sync to guest and only then run the command? Why not just execute using `ssh` command as we already do for example for `shell` preparation?

Reply:
That was my first intention, but IIRC it proved to be very tricky and unreliable to quote the commands from users:
  https://github.com/psss/tmt/pull/159/files/2e317f5014ce6851b3b9e9d24f7477da41e8b718#diff-4476c8abdfef907fe037b9351ba16b5cR205

But maybe I didn't try the `shlex`, variant. And there's one more reason- but let me get back to you on Friday, when I test it.